Detect inlining accurately and skip generating precode fixups for inlines that do not succeed

davidwrighton opened the pull request:
- Fix reporting of nested inlining to be correct
  - In particular, if a doubly nested inline succeeds, but the method that called it didn't succeed its inline, do not report that the inner nested inline succeeded
- Fix reporting of precode fixups to only be the precode fixups needed during actual use

davidwrighton committed:
Detect inlining accurately and skip generating precode fixups for inlines that do not succeed
- Fix reporting of nested inlining to be correct
  - In particular, if a doubly nested inline succeeds, but the method that called it didn't succeed its inline, do not report that the inner nested inline succeeded
- Fix reporting of precode fixups to only be the precode fixups needed during actual use

davidwrighton committed:
Fix jit formatting

davidwrighton commented:
@BruceForstall This is extracted from my rather larger crossgen2 cross module inlining work, and I expect it will cause the superpmi tasks to fail. Do you have any advice for me on how to run those tests manually?

davidwrighton committed:
Fix oops

BruceForstall commented:
Don't forget to change the JIT-EE GUID.

> @BruceForstall This is extracted from my rather larger crossgen2 cross module inlining work, and I expect it will cause the superpmi tasks to fail. Do you have any advice for me on how to run those tests manually?

Since you need to change the JIT-EE GUID, the spmi tasks are guaranteed to fail ("by design").

The JIT\superpmi\superpmicollect unit test does some basic SPMI testing on every PR.

For more extensive testing, I would suggest doing a PMI collection of Core_Root. e.g., with the current directory in the repo:
```
py src\coreclr\scripts\superpmi.py collect --pmi -assemblies c:\gh\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root -output_mch_path test.mch
```

You can replace `--pmi` with `--crossgen2` to do a collection from a crossgen2 run instead of a PMI run, which might be more appropriate, or you can just do both.

I'll try to review the SPMI code soon.

davidwrighton committed:
Update jit guid

BruceForstall commented on src/coreclr/tools/superpmi/superpmi-shim-collector/icorjitinfo.cpp:
```diff
@@ -115,6 +115,12 @@ CorInfoInline interceptor_ICJI::canInline(CORINFO_METHOD_HANDLE callerHnd,    /*
+void interceptor_ICJI::beginInlining(CORINFO_METHOD_HANDLE inlinerHnd,
+                                     CORINFO_METHOD_HANDLE inlineeHnd)
+{
+    // do nothing
```
```suggestion
    mc->cr->AddCall("beginInlining");
    original_ICorJitInfo->beginInlining(inlinerHnd, inlineeHnd);
```

Without this call, when doing a spmi collection (under crossgen), crossgen won't see this call from the JIT.

Normally, we'd also add `mc->cr->recBeginInlining()` and all the mechanism around saving/replaying/dumping this call. I guess it's ok we don't add that because nobody would use it -- on replay we don't need it.

davidwrighton replied:
Yes, this caused superpmi collections with crossgen2 to fail.

BruceForstall commented on src/coreclr/tools/superpmi/superpmi-shim-collector/icorjitinfo.cpp:
```diff
@@ -123,9 +129,12 @@ void interceptor_ICJI::reportInliningDecision(CORINFO_METHOD_HANDLE inlinerHnd,
-    mc->cr->AddCall("reportInliningDecision");
-    original_ICorJitInfo->reportInliningDecision(inlinerHnd, inlineeHnd, inlineResult, reason);
-    mc->cr->recReportInliningDecision(inlinerHnd, inlineeHnd, inlineResult, reason);
+    if (inlineResult == INLINE_PASS || inlineResult == INLINE_FAIL|| inlineResult == INLINE_NEVER)
+    {
```
Why not just drop this condition and always do the pass-through/recReportInliningDecision call? I don't think it would hurt.

davidwrighton replied:
Yep, seems to work.

BruceForstall commented on src/coreclr/tools/superpmi/superpmi/icorjitinfo.cpp:
```diff
@@ -108,8 +113,11 @@ void MyICJI::reportInliningDecision(CORINFO_METHOD_HANDLE inlinerHnd,
 {
-    jitInstance->mc->cr->AddCall("reportInliningDecision");
-    jitInstance->mc->cr->recReportInliningDecision(inlinerHnd, inlineeHnd, inlineResult, reason);
+    if (inlineResult == INLINE_PASS || inlineResult == INLINE_FAIL|| inlineResult == INLINE_NEVER)
+    {
```
Seems like we don't need this condition

davidwrighton replied:
Yep. I've removed it and things seem to work.

BruceForstall commented on src/coreclr/inc/corinfo.h:
```diff
@@ -954,11 +954,14 @@ enum CorInfoClassId
 {
-    INLINE_PASS                 = 0,    // Inlining OK
+    INLINE_PASS                     = 0,    // Inlining OK
+    INLINE_PREJIT_SUCCESS           = 1,
+    INLINE_CHECK_CAN_INLINE_SUCCESS = 2,
+    INLINE_CHECK_CAN_INLINE_VMFAIL  = 3,
```
IMO, it would be nice to have some comments for each of these values. E.g., I don't know how to read `INLINE_CHECK_CAN_INLINE_SUCCESS`.

davidwrighton replied:
Added.

BruceForstall reviewed (commented):
(trying to change from "request" to "comment")

BruceForstall commented:
@AndyAyersMS Inlining reporting changes

davidwrighton committed:
Address code review feedback, and fix SuperPMI detected JIT assertion around vector types

davidwrighton commented:
@BruceForstall I've addressed your concerns, and run SuperPMI. When looking at the failures I discovered that in the logs we were reporting an assertion during SuperPMI collection around compilation of System.Private.CoreLib. I've dropped a fix for that into this PR as well. The error message in the logs was:
```
[13:52:20] Assert failure(PID 97552 [0x00017d10], Thread: 5112 [0x13f8]): Assertion failed '(intrinsic == NI_Vector128_As) || (intrinsic == NI_Vector128_AsByte) || (intrinsic == NI_Vector128_AsDouble) || (intrinsic == NI_Vector128_AsInt16) || (intrinsic == NI_Vector128_AsInt32) || (intrinsic == NI_Vector128_AsInt64) || (intrinsic == NI_Vector128_AsSByte) || (intrinsic == NI_Vector128_AsSingle) || (intrinsic == NI_Vector128_AsUInt16) || (intrinsic == NI_Vector128_AsUInt32) || (intrinsic == NI_Vector128_AsUInt64) || (intrinsic == NI_Vector128_get_AllBitsSet) || (intrinsic == NI_Vector128_get_Count) || (intrinsic == NI_Vector128_get_Zero) || (intrinsic == NI_Vector128_GetElement) || (intrinsic == NI_Vector128_WithElement) || (intrinsic == NI_Vector128_ToScalar) || (intrinsic == NI_Vector128_ToVector256) || (intrinsic == NI_Vector128_ToVector256Unsafe) || (intrinsic == NI_Vector256_As) || (intrinsic == NI_Vector256_AsByte) || (intrinsic == NI_Vector256_AsDouble) || (intrinsic == NI_Vector256_AsInt16) || (intrinsic == NI_Vector256_AsInt32) || (intrinsic == NI_Vector256_AsInt64) || (intrinsic == NI_Vector256_AsSByte) || (intrinsic == NI_Vector256_AsSingle) || (intrinsic == NI_Vector256_AsUInt16) || (intrinsic == NI_Vector256_AsUInt32) || (intrinsic == NI_Vector256_AsUInt64) || (intrinsic == NI_Vector256_get_AllBitsSet) || (intrinsic == NI_Vector256_get_Count) || (intrinsic == NI_Vector256_get_Zero) || (intrinsic == NI_Vector256_GetElement) || (intrinsic == NI_Vector256_WithElement) || (intrinsic == NI_Vector256_GetLower) || (intrinsic == NI_Vector256_ToScalar)' in 'System.Runtime.Intrinsics.Vector256:Dot(System.Runtime.Intrinsics.Vector256`1[Vector`1],System.Runtime.Intrinsics.Vector256`1[Vector`1]):System.Numerics.Vector`1[Single]' during 'Importation' (IL size 58; hash 0x960b96cd; FullOpts)    File: C:\git3\runtime\src\coreclr\jit\hwintrinsic.cpp Line: 662    Image: C:\git3\runtime\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\corerun.exe
```

I was surprised that such a failure was only visible when looking into the logs, and did not cause the overall superpmi tool to tell me it failed somwhere.

AndyAyersMS commented:
I'm a bit surprised that the jit would report success to the VM for a "second level" inline. How did this happen?

You will probably want to modify `reportInliningDecision` in the runtime jit interface not to fire off ETW events for the new `CorInfoInline` values (VM fail is logged directly by the VM, and the check cases should be ignored).

I think there's enough context on the jit side to work up a simpler/more in-kind set of jit changes, but it's not super-important.

BruceForstall commented:
> I discovered that in the logs we were reporting an assertion during SuperPMI collection around compilation of System.Private.CoreLib

Did you fix this assert? It's been logged as https://github.com/dotnet/runtime/issues/65814 for a while, but nobody's investigated.

> I was surprised that such a failure was only visible when looking into the logs, and did not cause the overall superpmi tool to tell me it failed somwhere.

Ignoring failures in PMI runs is kind of by design -- there are lots of failures that can occur, and we don't want to "hard fail" on any of them -- we want to do "as much as possible". In the case of SuperPMI collection, we still get useful artifacts by collecting as much as possible and ignoring failures. It would be good to at least surface failures so we try to do something about them -- at least the VM/JIT asserts.

BruceForstall commented on src/coreclr/jit/hwintrinsic.cpp:
```diff
@@ -640,48 +640,15 @@ static bool isSupportedBaseType(NamedIntrinsic intrinsic, CorInfoType baseJitTyp
-           (intrinsic == NI_Vector256_get_Zero) || (intrinsic == NI_Vector256_GetElement) ||
-           (intrinsic == NI_Vector256_WithElement) || (intrinsic == NI_Vector256_GetLower) ||
-           (intrinsic == NI_Vector256_ToScalar));
+    assert(isa == InstructionSet_Vector256 || InstructionSet_Vector128);
```
```suggestion
    assert((isa == InstructionSet_Vector256) || (isa == InstructionSet_Vector128));
```
The way you wrote it, I think it would never fail.

I'm not sure what the difference is between the two asserts, but it seems reasonable to me. @tannergooding any comment?

BruceForstall commented on src/coreclr/jit/hwintrinsic.cpp:
```diff
@@ -640,48 +640,15 @@ static bool isSupportedBaseType(NamedIntrinsic intrinsic, CorInfoType baseJitTyp
-           (intrinsic == NI_Vector128_GetElement) || (intrinsic == NI_Vector128_GetLower) ||
-           (intrinsic == NI_Vector128_GetUpper) || (intrinsic == NI_Vector128_ToScalar) ||
-           (intrinsic == NI_Vector128_WithElement));
+    assert(isa == InstructionSet_Vector64 || InstructionSet_Vector128);
```
```suggestion
    assert((isa == InstructionSet_Vector64) || (isa == InstructionSet_Vector128));
```

davidwrighton committed:
Address code review feedback

BruceForstall reviewed (approved):
The SPMI changes look good.

I'd like @tannergooding to comment on the changed assert (if it's doing the "right thing", then great!). Also, @AndyAyersMS should sign off on the inliner parts.

tannergooding commented on src/coreclr/jit/hwintrinsic.cpp:
```diff
@@ -640,48 +640,15 @@ static bool isSupportedBaseType(NamedIntrinsic intrinsic, CorInfoType baseJitTyp
-           (intrinsic == NI_Vector256_get_Zero) || (intrinsic == NI_Vector256_GetElement) ||
-           (intrinsic == NI_Vector256_WithElement) || (intrinsic == NI_Vector256_GetLower) ||
-           (intrinsic == NI_Vector256_ToScalar));
+    assert((isa == InstructionSet_Vector256) || (isa == InstructionSet_Vector128));
```
I'm fine with reducing this down.

The intent here was to intentionally assert coverage of known generic methods (that is methods that explicitly take <T> rather than a concrete type like int).

However, I don't think it was actually buying us much, so I'm fine reducing this down to what you have

tannergooding replied:
Noting that such generic methods should only exist in the `Vector64/128/256` ISAs, never in the other ISAs, so this remains "correct".

davidwrighton commented:
The SuperPMI failures in here are due to the presence of a jit interface version guid change, nothing else.

tannergooding commented:
https://github.com/dotnet/runtime/issues/69197 tracks that issue, just as an FYI
